app/vmselect/promql/rollup.go

```diff
@@ -900,17 +900,26 @@ func getMaxPrevInterval(scrapeInterval int64) int64 {
 	return scrapeInterval + scrapeInterval/8
 }
 
-func removeCounterResets(values []float64) {
+func removeCounterResets(values []float64, timestamps []int64, maxStalenessInterval int64) {
 	// There is no need in handling NaNs here, since they are impossible
 	// on values from vmstorage.
 	if len(values) == 0 {
 		return
 	}
 	var correction float64
 	prevValue := values[0]
+	prevTS := timestamps[0]
 	for i, v := range values {
 		d := v - prevValue
 		if d < 0 {
+			currTS := timestamps[i]
+			if maxStalenessInterval > 0 && ((currTS - prevTS) > maxStalenessInterval) {
+				// don't apply correction if previous sample exceeds staleness interval
+				correction = 0
+				prevValue = v
+				prevTS = timestamps[i]
```

Comment:
It is better from readability and performance PoV to use `timestamps[i-1]` instead of `prevTS`. I'm this case the code, which accesses the previous timestamp, will be localized in smaller number of lines.